src/Storages/HDFS/HDFSCommon.cpp

```diff
@@ -161,46 +142,186 @@ HDFSBuilderWrapper createHDFSBuilder(const String & uri_str, const Poco::Util::A
         else
             user = user_info;
 
-        hdfsBuilderSetUserName(builder.get(), user.c_str());
+        hdfsBuilderSetUserName(get(), user.c_str());
     }
 
-    hdfsBuilderSetNameNode(builder.get(), host.c_str());
+    hdfsBuilderSetNameNode(get(), host.c_str());
     if (port != 0)
     {
-        hdfsBuilderSetNameNodePort(builder.get(), port);
+        hdfsBuilderSetNameNodePort(get(), port);
     }
 
     if (config.has(HDFSBuilderWrapper::CONFIG_PREFIX))
     {
-        builder.loadFromConfig(config, HDFSBuilderWrapper::CONFIG_PREFIX);
+        loadFromConfig(HDFSBuilderWrapper::CONFIG_PREFIX);
     }
 
     if (!user.empty())
     {
         String user_config_prefix = HDFSBuilderWrapper::CONFIG_PREFIX + "_" + user;
         if (config.has(user_config_prefix))
         {
-            builder.loadFromConfig(config, user_config_prefix, true);
+            loadFromConfig(user_config_prefix, true);
+        }
+    }
+
+    if (need_kinit)
+    {
+        runKinit();
+    }
+}
+
+HDFSFSPool::HDFSFSPool(uint32_t max_items_, uint32_t min_items_, HDFSBuilderWrapperPtr builder_)
+    : max_items(max_items_), min_items(std::min(max_items_, min_items_)), builder(std::move(builder_)), current_index(0)
+{
+    pool.reserve(max_items);
+
+    if (min_items)
+    {
+        pool.resize(min_items);
+        ThreadPool thread_pool(std::min(static_cast<uint32_t>(32), min_items));
+        for (size_t i = 0; i < min_items; ++i)
+        {
+            thread_pool.scheduleOrThrowOnError([this, i]()
+            {
+                pool[i] = createHDFSFS(builder->get());
+            });
+        }
+        thread_pool.wait();
+    }
+}
+
+HDFSFSPtr HDFSFSPool::getFS()
+{
+    std::lock_guard lock{mutex};
+    return unsafeGetFS();
+}
+
+HDFSFSPtr HDFSFSPool::unsafeGetFS()
+{
+    if (current_index == pool.size())
+    {
+        if (pool.size() == max_items)
+        {
+            current_index = 0;
+        }
+        else
+        {
+            pool.emplace_back(createHDFSFS(builder->get()));
+        }
+    }
+    return pool[current_index++];
+}
+
+bool HDFSFSPool::tryCallFS(HDFSFSPtr & fs, std::function<bool(HDFSFSPtr &)> callback)
+{
+    int retry = 0;
+    while (retry++ < 3)
+    {
+        if (callback(fs))
+            return true;
+
+        {
+            std::lock_guard lock{mutex};
+
+            bool still_in_pool = false;
+            for (auto & i : pool)
+            {
+                if (i == fs)
```

Comment:
isn't it more correct to compare `i.get()` and `fs.get()`?

Reply:
Don't worry. It's implemented in 
```
template<class _Tp, class _Up>
inline _LIBCPP_INLINE_VISIBILITY
bool
operator==(const shared_ptr<_Tp>& __x, const shared_ptr<_Up>& __y) _NOEXCEPT
{
    return __x.get() == __y.get();
}
```